File: ex/lib/gralkor/client/http.ex
defmodule Gralkor.Client.HTTP do
  @moduledoc """
  Real `Gralkor.Client` implementation over HTTP using `Req`.

  Reads config from `Application.get_env(:gralkor_ex, :client_http)`:

    * `:url` — required. Base URL of the Gralkor server (e.g.
      `"http://127.0.0.1:4000"`).
    * `:plug` — optional `Req.Test` plug tuple for stubbing in tests.
      Unset in production so Req hits the network directly.

  No auth: Gralkor is expected to run under the consumer app's
  supervision tree, bound to loopback. The consumer owns the trust
  boundary.

  Per-endpoint `receive_timeout`s, calibrated to the workload. 429 retry
  ownership for Vertex-upstream rate-limits lives inside `/recall` on
  the server — see `gralkor/TEST_TREES.md > Retry ownership`. No layer
  above the server retries this class. Under sustained Vertex throttling
  `/recall` returns 429 (for exhausted retries) or 504 (for deadline);
  the consumer's `jido_gralkor` plugin then degrades gracefully to a
  memory-less turn.

    * `/health` (2 s) — cheap liveness check; tight so `Gralkor.Connection`
      doesn't flap when the server is under LLM load.
    * `/recall` (12 s) — matches the server's `/recall` deadline. The
      server body runs graph search (`graphiti.search()` — RRF, edges
      only, calls the embedder) plus `interpret_facts`, with one 429
      retry absorbed internally inside the 12 s budget. Tight — a
      server-side 504 may race the transport; revisit if it bites.
    * `/tools/memory_search` (30 s) — *slow* graph search
      (`graphiti.search_()` with `COMBINED_HYBRID_SEARCH_CROSS_ENCODER`
      — cross-encoder reranking + BFS) plus `interpret_facts`. More
      upstream work per call than `/recall`; sized a few seconds higher.
    * `/capture` (5 s) — server returns 204 immediately after buffering.
      No synchronous LLM call here; the flush runs in the server-side
      capture buffer (its own retry schedule).
    * `/session_end` (5 s) — server returns 204 immediately after
      scheduling the flush.
    * `/tools/memory_add` (60 s) — Graphiti entity/edge extraction is
      slow; only reached from a background `Task` in the consumer, so
      the agent never waits.
    * `/build-indices`, `/build-communities` (`:infinity`) — admin
      operations that scan the whole graph; can run for minutes to
      hours on a populated database. The operator invokes them
      explicitly, so blocking the caller is fine.

  Returns `{:error, reason}` on non-2xx or transport failure; raises on
  missing config or blank session_id. Callers let those surface.
  """

  @behaviour Gralkor.Client

  @health_timeout_ms 2_000
  @capture_timeout_ms 5_000
  @end_session_timeout_ms 5_000
  @recall_timeout_ms 12_000
  @tool_search_timeout_ms 30_000
  @memory_add_timeout_ms 60_000


  @impl true
  def recall(group_id, session_id, query) do
    body = %{group_id: group_id, query: query}
    body = if is_binary(session_id), do: Map.put(body, :session_id, session_id), else: body

    post("/recall", body, @recall_timeout_ms)
    |> case do
      {:ok, %{"memory_block" => ""}} -> {:ok, nil}
      {:ok, %{"memory_block" => block}} -> {:ok, block}
      {:ok, body} -> {:error, {:unexpected_body, body}}
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def capture(session_id, group_id, messages) when is_list(messages) do
    require_session_id!(session_id)

    post(
      "/capture",
      %{
        session_id: session_id,
        group_id: group_id,
        messages: Enum.map(messages, &message_to_json/1)
      },
      @capture_timeout_ms
    )
    |> case do
      {:ok, _body} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  defp message_to_json(%Gralkor.Message{role: role, content: content}) do
    %{role: role, content: content}
  end

  @impl true
  def memory_search(group_id, session_id, query) do
    require_session_id!(session_id)

    post(
      "/tools/memory_search",
      %{group_id: group_id, session_id: session_id, query: query},
      @tool_search_timeout_ms
    )
    |> case do
      {:ok, %{"text" => text}} -> {:ok, text}
      {:ok, body} -> {:error, {:unexpected_body, body}}
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def end_session(session_id) do
    require_session_id!(session_id)

    case post("/session_end", %{session_id: session_id}, @end_session_timeout_ms) do
      {:ok, _body} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def memory_add(group_id, content, source_description) do
    body = %{group_id: group_id, content: content}

    body =
      if is_binary(source_description),
        do: Map.put(body, :source_description, source_description),
        else: body

    case post("/tools/memory_add", body, @memory_add_timeout_ms) do
      {:ok, _body} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def health_check do
    {url, req_opts} = req_options(@health_timeout_ms)

    case Req.get(url <> "/health", req_opts) |> handle_response() do
      {:ok, _body} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def build_indices do
    case post("/build-indices", %{}, :infinity) do
      {:ok, %{"status" => status}} when is_binary(status) -> {:ok, %{status: status}}
      {:ok, body} -> {:error, {:unexpected_body, body}}
      {:error, reason} -> {:error, reason}
    end
  end

  @impl true
  def build_communities(group_id) when is_binary(group_id) do
    case post("/build-communities", %{group_id: group_id}, :infinity) do
      {:ok, %{"communities" => c, "edges" => e}} when is_integer(c) and is_integer(e) ->
        {:ok, %{communities: c, edges: e}}

      {:ok, body} ->
        {:error, {:unexpected_body, body}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp require_session_id!(id) when is_binary(id) and id != "", do: :ok
  defp require_session_id!(_), do: raise(ArgumentError, "session_id must be a non-blank string")

  defp post(path, body, timeout_ms) do
    {url, req_opts} = req_options(timeout_ms)
    opts = [json: body] ++ req_opts

    Req.post(url <> path, opts) |> handle_response()
  end

  defp req_options(timeout_ms) do
    config = Application.fetch_env!(:gralkor_ex, :client_http)
    url = Keyword.fetch!(config, :url)

    req_opts =
      [
        receive_timeout: timeout_ms,
        retry: false
      ]
      |> maybe_put(:plug, Keyword.get(config, :plug))

    {url, req_opts}
  end

  defp maybe_put(opts, _key, nil), do: opts
  defp maybe_put(opts, key, value), do: Keyword.put(opts, key, value)

  defp handle_response({:ok, %Req.Response{status: status, body: body}}) when status in 200..299,
    do: {:ok, body}

  defp handle_response({:ok, %Req.Response{status: status, body: body}}),
    do: {:error, {:http_status, status, body}}

  defp handle_response({:error, reason}), do: {:error, reason}
end

File: ts/src/client/http.ts
import type { GralkorClient, Message, Result } from "../client.js";

export interface GralkorHttpClientOptions {
  /** Base URL of the Gralkor server (e.g. `http://127.0.0.1:4000`). No trailing slash required. */
  baseUrl: string;
  /** Override fetch, e.g. for tests. Defaults to global fetch. */
  fetch?: typeof fetch;
}

/**
 * HTTP adapter for {@link GralkorClient}.
 *
 * No auth: the server binds to loopback and expects its consumer to supervise it.
 * No retries at this layer: non-2xx responses and transport errors surface immediately.
 * See `gralkor/TEST_TREES.md` > Retry ownership — 429 retry lives inside the
 * server's `/recall` handler; no layer above it retries this class.
 *
 * Per-endpoint timeouts (milliseconds), calibrated to the workload:
 *
 *   - `/health`                — 2 000
 *   - `/recall`                — 12 000 (matches the server's `/recall` deadline; tight — a server 504 may race the transport, revisit if it bites)
 *   - `/capture`               — 5 000
 *   - `/session_end`           — 5 000
 *   - `/tools/memory_search`   — 30 000
 *   - `/tools/memory_add`      — 60 000 (Graphiti extraction is slow)
 *   - `/build-indices`         — none (admin; minutes-to-hours on large graphs)
 *   - `/build-communities`     — none (admin; minutes-to-hours on large graphs)
 *
 * Blank session ids throw `Error` — Gralkor requires a non-blank session_id.
 */
export class GralkorHttpClient implements GralkorClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: GralkorHttpClientOptions) {
    if (!options.baseUrl) throw new Error("baseUrl is required");
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.fetchImpl = options.fetch ?? fetch;
  }

  async recall(
    groupId: string,
    sessionId: string | null,
    query: string,
    maxResults?: number,
  ): Promise<Result<string | null>> {
    const body: Record<string, unknown> = { group_id: groupId, query };
    if (typeof sessionId === "string") body.session_id = sessionId;
    if (maxResults !== undefined) body.max_results = maxResults;
    const res = await this.post("/recall", body, 12_000);
    if ("error" in res) return res;
    const respBody = res.ok as { memory_block?: string };
    if (respBody.memory_block === undefined) return { error: { kind: "unexpected_body", body: respBody } };
    if (respBody.memory_block === "") return { ok: null };
    return { ok: respBody.memory_block };
  }

  async capture(sessionId: string, groupId: string, messages: Message[]): Promise<Result<true>> {
    requireSessionId(sessionId);
    const res = await this.post(
      "/capture",
      { session_id: sessionId, group_id: groupId, messages },
      5_000,
    );
    return "error" in res ? res : { ok: true };
  }

  async endSession(sessionId: string): Promise<Result<true>> {
    requireSessionId(sessionId);
    const res = await this.post("/session_end", { session_id: sessionId }, 5_000);
    return "error" in res ? res : { ok: true };
  }

  async memorySearch(
    groupId: string,
    sessionId: string,
    query: string,
    maxResults?: number,
    maxEntityResults?: number,
  ): Promise<Result<string>> {
    requireSessionId(sessionId);
    const body: Record<string, unknown> = { group_id: groupId, session_id: sessionId, query };
    if (maxResults !== undefined) body.max_results = maxResults;
    if (maxEntityResults !== undefined) body.max_entity_results = maxEntityResults;
    const res = await this.post("/tools/memory_search", body, 30_000);
    if ("error" in res) return res;
    const respBody = res.ok as { text?: string };
    if (respBody.text === undefined) return { error: { kind: "unexpected_body", body: respBody } };
    return { ok: respBody.text };
  }

  async memoryAdd(
    groupId: string,
    content: string,
    sourceDescription: string | null,
  ): Promise<Result<true>> {
    const body: Record<string, unknown> = { group_id: groupId, content };
    if (sourceDescription !== null) body.source_description = sourceDescription;
    const res = await this.post("/tools/memory_add", body, 60_000);
    return "error" in res ? res : { ok: true };
  }

  async healthCheck(): Promise<Result<true>> {
    const res = await this.request("GET", "/health", undefined, 2_000);
    return "error" in res ? res : { ok: true };
  }

  async buildIndices(): Promise<Result<{ status: string }>> {
    const res = await this.post("/build-indices", {}, undefined);
    if ("error" in res) return res;
    const body = res.ok as { status?: string };
    if (typeof body.status !== "string") return { error: { kind: "unexpected_body", body } };
    return { ok: { status: body.status } };
  }

  async buildCommunities(
    groupId: string,
  ): Promise<Result<{ communities: number; edges: number }>> {
    const res = await this.post("/build-communities", { group_id: groupId }, undefined);
    if ("error" in res) return res;
    const body = res.ok as { communities?: number; edges?: number };
    if (typeof body.communities !== "number" || typeof body.edges !== "number") {
      return { error: { kind: "unexpected_body", body } };
    }
    return { ok: { communities: body.communities, edges: body.edges } };
  }

  private post(path: string, body: unknown, timeoutMs: number | undefined): Promise<Result<unknown>> {
    return this.request("POST", path, body, timeoutMs);
  }

  private async request(
    method: "GET" | "POST",
    path: string,
    body: unknown | undefined,
    timeoutMs: number | undefined,
  ): Promise<Result<unknown>> {
    const controller = new AbortController();
    const timer =
      timeoutMs === undefined ? null : setTimeout(() => controller.abort(), timeoutMs);

    try {
      const res = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers: body !== undefined ? { "content-type": "application/json" } : undefined,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      if (res.status >= 200 && res.status < 300) {
        const text = await res.text();
        if (text === "") return { ok: null };
        try {
          return { ok: JSON.parse(text) };
        } catch {
          return { ok: text };
        }
      }

      const errBody = await res.text().catch(() => "");
      return { error: { kind: "http_status", status: res.status, body: errBody } };
    } catch (err) {
      return {
        error: {
          kind: "network",
          cause: err instanceof Error ? err.message : String(err),
        },
      };
    } finally {
      if (timer !== null) clearTimeout(timer);
    }
  }
}

function requireSessionId(id: string): void {
  if (typeof id !== "string" || id === "") {
    throw new Error("session_id must be a non-blank string");
  }
}
